python/src/trezorlib/cli/device.py

```diff
@@ -334,17 +339,278 @@ def set_busy(
+      pip3 install cryptography
+    """
     if hex_challenge is None:
         hex_challenge = secrets.token_hex(32)
-    click.echo(f"Challenge: {hex_challenge}")
+
     challenge = bytes.fromhex(hex_challenge)
     msg = device.authenticate(client, challenge)
-    click.echo(f"Signature of challenge: {msg.signature.hex()}")
-    click.echo(f"Device certificate: {msg.certificates[0].hex()}")
-    for cert in msg.certificates[1:]:
-        click.echo(f"CA certificate: {cert.hex()}")
+
+    if raw:
+        click.echo(f"Challenge: {hex_challenge}")
+        click.echo(f"Signature of challenge: {msg.signature.hex()}")
+        click.echo(f"Device certificate: {msg.certificates[0].hex()}")
+        for cert in msg.certificates[1:]:
+            click.echo(f"CA certificate: {cert.hex()}")
+        return
+
+    _require_cryptography()
+
+    worst_level = DEBUG
+
+    def print_step(level: int, text: str) -> None:
+        nonlocal worst_level
+
+        worst_level = max(worst_level, level)
+        if level == ERROR:
+            level_str = click.style("ERROR", fg="red")
+        elif level == WARNING:
+            level_str = click.style("WARNING", fg="yellow")
+        elif level == INFO:
+            level_str = click.style("INFO", fg="blue")
+        elif level == DEBUG:
+            level_str = click.style("OK", fg="green")
+        else:
+            raise RuntimeError("Invalid log level")
+
+        click.echo(f"[{level_str}] {text}")
+
+    from cryptography import exceptions  # noqa: I900
+
+    CHALLENGE_HEADER = b"AuthenticateDevice:"
+    challenge_bytes = (
+        len(CHALLENGE_HEADER).to_bytes(1, "big")
+        + CHALLENGE_HEADER
+        + len(challenge).to_bytes(1, "big")
+        + challenge
+    )
+
+    try:
+        first_cert = Certificate(msg.certificates[0])
+    except Exception:
+        print_step(ERROR, "Failed to parse device certificate.")
+        sys.exit(5)
+
+    try:
+        first_cert.verify(msg.signature, challenge_bytes)
+    except exceptions.InvalidSignature:
+        print_step(ERROR, "Challenge verification failed.")
+    else:
+        print_step(DEBUG, "Challenge verified successfully.")
+
+    for issuer in msg.certificates[1:]:
+        try:
+            cert = Certificate(issuer)
+        except Exception:
+            print_step(ERROR, "Failed to parse CA certificate.")
+            continue
+
+        if skip_whitelist:
+            print_step(INFO, "Skipping public key whitelist check.")
+        else:
+            if cert.public_key() not in TS3_CA_WHITELIST:
+                print_step(WARNING, f"CA certificate not in whitelist: {cert}")
+
+        try:
+            first_cert.verify_issued_by(cert)
+        except exceptions.InvalidSignature:
+            print_step(ERROR, f"Certificate verification failed: {cert}")
+            continue
+        else:
+            print_step(DEBUG, f"Certificate verified successfully: {cert}")
```

Comment:
In both cases above, semantically it would make more sense to me to see the certificate subject name that was verified successfully or failed verification, not the issuer name. In other words who was verified successfully, not by whom it was verified.